Refactor/messages crate

bobozaur opened the pull request:
Refactor of messages crate.

Planned steps:
  - Initially, development will be carried in a new crate, `messages2`.
  - After the crate is complete, it's integration will be done within `aries_vcx`.

Patrik-Stas reviewed (commented):
Barely went through everything but went through a little bit. Produced this small PR in the process
https://github.com/hyperledger/aries-vcx/pull/755

Patrik-Stas commented on messages2/src/a2a.rs:
```diff
@@ -0,0 +1,190 @@
+    {
+        use serde::__private::de::{ContentDeserializer, TaggedContentVisitor};
+
+        let tag_visitor = TaggedContentVisitor::<MessageType>::new("@type", "internally tagged enum A2AMessage");
```
perhaps you could explain not only why it's okay to use `TaggedContentVisitor`, but what it's doing - it's not easily lookup-able since it's not yet public API. Same would go for `ContentDeserializer`

bobozaur replied:
Done.

Patrik-Stas commented:
I'd suggest to take a look at some instance where breaking change protocol actually happened, like https://github.com/hyperledger/aries-rfcs/blob/main/features/0454-present-proof-v2/README.md 
and try to add support for those messages. I think this would be good practical test as to how well the created protocol versioning abstractions, mechanisms hold.

Patrik-Stas commented on messages2/src/a2a.rs:
```diff
@@ -0,0 +1,190 @@
+impl DelayedSerde for A2AMessage {
+    type Seg = MessageFamily;
+
+    fn delayed_deserialize<'de, D>(seg: Self::Seg, deserializer: D) -> Result<Self, D::Error>
```
though you already touched down on today's call, I think it'd be nice to write brief explanation here as to what is going on. How is this delaying serialization and why is it more optimal?

Patrik-Stas commented:
In PR #755 I added some tests, how would I go about it in the opposite direction on Message Family level. Eg something like this 
```
#[test]
    fn connections_to_strings() {
        let msg_type = MessageType::new(Prefix::DidCommOrg, ConnectionV1_0::Invitation.into());
        assert_eq!(msg_type.to_string(), "https://didcomm.org/connections/1.0/invitation");
    }
 ```

Obviously I need to implement `Display` trait and this would be handy for logging in general. But I am not sure how go about implementation - surely we could reuse `impl Serialize for MessageType` to generate the final string for `Display` implementation?

bobozaur commented on messages2/src/message_type/message_family/cred_issuance.rs:
```diff
@@ -0,0 +1,80 @@
+}
+
+#[derive(Copy, Clone, Debug, AsRefStr, EnumString, PartialEq)]
+#[strum(serialize_all = "kebab-case")]
```
This is what's used predominantly across the ecosystem, although it seems `snake_case` is used in some places too. So I'm basically resorting to `kebab-case` as a default.

Patrik-Stas replied:
ah yeah ofc, got it confused with rust conventions

bobozaur commented:
> I'd suggest to take a look at some instance where breaking change protocol actually happened, like https://github.com/hyperledger/aries-rfcs/blob/main/features/0454-present-proof-v2/README.md and try to add support for those messages. I think this would be good practical test as to how well the created protocol versioning abstractions, mechanisms hold.

To accommodate this in the future, we would expand the `PresentProof` message family and add a `V2` variant and it's minor version variants. 

Then create the "end messages" (e.g.: PresentProofV2_0), implement `ConcreteMessage` on them, create enum variants in the `A2AMessage::PresentProof` with `V2` and minor versions as well and the rest will pretty much take care of itself.

bobozaur commented:
> In PR #755 I added some tests, how would I go about it in the opposite direction on Message Family level. Eg something like this
> 
> ```
> #[test]
>     fn connections_to_strings() {
>         let msg_type = MessageType::new(Prefix::DidCommOrg, ConnectionV1_0::Invitation.into());
>         assert_eq!(msg_type.to_string(), "https://didcomm.org/connections/1.0/invitation");
>     }
> ```
> 
> Obviously I need to implement `Display` trait and this would be handy for logging in general. But I am not sure how go about implementation - surely we could reuse `impl Serialize for MessageType` to generate the final string for `Display` implementation?

While using the `Serialize` impl would work, it's much faster to bypass the entire `serde_json` machinery and construct the message type directly.

I took the liberty of introducing `MessageFamily::as_parts()` which provide the constants declared within the types. The constants, along with the prefix, are then written to the string formatter in the `Display` impl for `MessageType`.

I also adjusted your tests to rather use `MessageType::to_string()` against `&str` types.

bobozaur commented:
# Combine MD and FD
Given the community call discussion, it seems like a good idea to actually combine the message decorators and field decorators into a structure defined for each message. The whole point is to destructure it anyway, so that's most likely fine. 

Our goal is to mainly keep the protocol state machine relevant data separate from the protocol state machine agnostic data, so we can process them independently.

# Field Decorators in nested JSON
Per the [RFC](https://github.com/hyperledger/aries-rfcs/blob/main/concepts/0011-decorators/README.md#decorator-scope), a *field decorator* is used as a sibling to the field it decorates. So they technically can be found at lower levels of the JSON.

I don't know of any actual implementations of something like that (apart from nested messages as with the OOB connectionless stuff, but that's handled as a contained message). 

If, however, something like that surfaces, I think we can easily accommodate it to the same message structure that we have, with the caveat of doing some conversions through some intermediary types to normalize the data.

bobozaur commented:
Finally, the `messages2` crate is feature complete. Here's a list of things remaining before it can be merged and used in the other parts of the workspace:

- Move modules around:
	- Move things used by messages_macros crate somewhere where they can be used with fully qualified paths
	
- [ ] Documentation:
	- [ ] Document messages with links to their respective RFCs
	- [ ] Document more intricate methods/types and explain what they do and how

- [ ] Testing:
	- [ ] Test deserialization/serialization of every message and their type
	- [ ] Test protocol registry minor version resolution

- [ ] Renames:
	- [ ] Find better suited names for certain types/methods

This list is *not* exhaustive. Ideally, as progress is made or ideas are proposed, specific items would be added to it, such as:
- *Rename `This` to `Other`*
- *Add tests for serialization/deserialization of `Something`*

Any help in identifying, adding or even implementing items on this list is much appreciated. Reviewing the PR should result in ideas of what would be needed.

In terms of documentation, if it's not easily understandable in a short time (10-15 seconds) what a piece of code does and how, then that's a good sign that comments would be needed. I personally prefer exhaustive comments in favor of little/no comments.

Proposing items can be done by commenting on this PR or reaching out directly. I will then add them to the list. There's **no** bad idea of things that should be documented/renamed/tested/etc., so don't be shy.

Patrik-Stas commented on messages2/src/decorators/thread.rs:
```diff
@@ -0,0 +1,86 @@
+
+use serde::{Deserialize, Serialize, Serializer};
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
```
I think this can be just `Eq` (i know it was `PartialEq` in original messages crate, but see no reason no to make it `Eq`, or am I missing something?)

bobozaur replied:
Why would we implement `Eq`? The `PartialEq` is mostly for tests really, so we can easily do assertions.

`Eq` is a marker trait that essentially says "this type has some additional properties" ([docs](https://doc.rust-lang.org/beta/std/cmp/trait.Eq.html)).

Most of the time `Eq` is not needed, unless you're implementing some sort of primitive and need to consider the properties outlined by the trait.

Patrik-Stas replied:
It's not needed, but it just seem reasonable that message equality is reflexive, symmetrical and transitive. I can't imagine why that would not hold. 
`Eq` implies `PartialEq`, but not the other way around, so the fact that only `PartialEq` raises in me a question: "eh does equivalency not hold here? Why would that be so?"

Not a big deal though, consider this and feel free to mark this conversation as Resolved.

bobozaur replied:
I feel it's more of a YAGNI thing.

mirgee commented on messages2/src/protocols/cred_issuance/mod.rs:
```diff
@@ -0,0 +1,170 @@
+#[derive(Copy, Clone, Debug, PartialEq)]
+struct CredentialPreviewMsgType;
+
+impl From<CredentialPreviewMsgType> for CredentialIssuanceV1_0Kind {
```
What is the purpose of this impl? Seems unused.

bobozaur replied:
It is. Looks like a leftover from before the `MessageType` refactoring into `Protocol` and the other types representing the message kind. Great catch!

In fact, this gives me an idea that could make serialization/deserialization easier for these types (that don't represent a standalone message type).

bobozaur replied:
Simplified deserialization for these *non-standalone* message types. The `From` impls stayed as a result.

bobozaur commented on messages2/src/protocols/out_of_band/invitation.rs:
```diff
@@ -0,0 +1,52 @@
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub accept: Option<Vec<MimeType>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub handshake_protocols: Option<Vec<String>>, // TODO: Protocol Registry
```
Ah, yeah. This should be a protocol as well.

bobozaur replied:
Fixed.

bobozaur replied:
Fixed.

Patrik-Stas reviewed (changes requested):
Good job! The 
```
Message<Content, Decorator>
```
and the related traits, the `MessageWithKind` blanket impl etc. seem to play pretty well together. 

Just please remove reformatting changes applied across the board in unrelated crates. (it also makes github really slow, making it bit more difficult to review, as there's many modified files)

Some of the unfinished pieces which are not too critical good be very good first issues for newcomers! Leave them something! :-) This is good and well organized code to learn from. I left suggestions to do this with some of my review comments.

bobozaur commented on messages2/src/msg_types/types/traits.rs:
```diff
@@ -0,0 +1,36 @@
+
+    fn try_resolve_version(minor: u8) -> MsgTypeResult<Self>;
+
+    fn as_version_parts(&self) -> (u8, u8);
```
Major and minor versions, the `version_parts`. It is used by the `MessageType` derive macro.

bobozaur replied:
Documented the module.

Patrik-Stas replied:
I see, it's updated to 
```
    /// Returns the major and minor versions of the protocol.
    fn as_version_parts(&self) -> (u8, u8);
```
What is confusing is that the trait is named `MajorVersion` but `as_version_parts` returns both major and minor version.

Patrik-Stas replied:
I see, yeah I think rename to `ProtocolVersion` would clear that up

bobozaur commented on messages2/src/msg_types/types/traits.rs:
```diff
@@ -0,0 +1,36 @@
+
+    const MAJOR: u8;
+
+    fn try_resolve_version(minor: u8) -> MsgTypeResult<Self>;
```
These are used in the `MessageType` derive macro.
Documented the module.

bobozaur commented on messages2/src/msg_types/types/traits.rs:
```diff
@@ -0,0 +1,36 @@
+    }
+}
+
+pub trait ProtocolName: Sized {
```
Used in the `MessageType` derive macro.
Documented the module.

bobozaur commented on messages2/src/misc/mod.rs:
```diff
@@ -0,0 +1,82 @@
+        assert_eq!(expected, deserialized)
+    }
+
+    pub fn test_msg<V, T, U>(content: T, decorators: U, mut json: Value)
```
Fair enough.

bobozaur replied:
Done.

bobozaur commented on messages2/src/msg_types/types/present_proof.rs:
```diff
@@ -0,0 +1,92 @@
+
+    #[test]
+    fn test_msg_type_presentation() {
+        test_utils::test_msg_type(PROTOCOL, KIND_PRESENTATION, PresentProofV1::V1_0(PhantomData))
```
Fair. Will inline the consts.

bobozaur replied:
Or maybe leave this as a good first issue?

bobozaur replied:
Well, I did end up doing this so we could have serialization test coverage too.

bobozaur commented on messages2/src/protocols/nameless/report_problem.rs:
```diff
@@ -0,0 +1,164 @@
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Where {
```
Ah, I missed this one through all the tests adjustments. Will improve the extended test.

bobozaur commented on messages2/src/protocols/mod.rs:
```diff
@@ -0,0 +1,2 @@
+pub mod traits;
+pub mod nameless;
```
`payload` is a decent name I think. It would be great if we could name both `crate::msg_types::types` and `crate::protocols::nameless` the same. What do you think about something like "data_impls"?

bobozaur replied:
Renamed modules so the directory structure now looks like:
- src
    - msg_types
        - protocols
    - msg_fields
        - protocols

bobozaur commented on messages2/src/message.rs:
```diff
@@ -0,0 +1,63 @@
+/// instrumental to the message processing, not an appendix to the message (such as `~thread` or
+/// `~timing`).
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct Message<C, D = Nothing> {
```
Yeah, that would be nice. Maybe rename `AriesMessage` to `ProtocolMessage` ? Or, or, or... rename `Message` to `MsgParts`?

bobozaur commented on messages2/src/decorators/timing.rs:
Very good catch. It slipped by me after first drafting the message.
I did find a better approach tho, based on the impl of `DateTime::to_rfc3339_opts`. (our version doesn't allocate a String just to drop it after reading and serializing the bytes in the serialization buffer)

Patrik-Stas commented on messages2/src/protocols/nameless/basic_message.rs:
```diff
@@ -0,0 +1,87 @@
+pub type BasicMessage = Message<BasicMessageContent, BasicMessageDecorators>;
+
+#[derive(Clone, Debug, Deserialize, Serialize, MessageContent, PartialEq)]
+#[message(kind = "BasicMessageV1_0::Message")]
```
fair enough :-)

bobozaur commented on messages2/src/decorators/attachment.rs:
```diff
@@ -0,0 +1,211 @@
+        let data = "test_base64_str".to_owned();
+
+        let json = json!({ "base64": data });
+        let attach_type = AttachmentType::Base64(data);
```
Replaced `String` with `Vec<u8>`

Patrik-Stas replied:
Well, the most safe, efficient and convenient solution would be to do the encoding upon setting it. That way messages would assure the encoding is correct without having to validate already encoded b64. Also convenient for user. 
We can leave that for later, maybe good first issue :-)

bobozaur replied:
We could, but that means locking all consumers into the same base64 encoding library. I think it's not a huge thing to allow users to pass arbitrary input as long as they realize that the other party won't understand it unless it is base64 encoded.

It would also be possible in some cases for users to already have base64 encoded data, so encoding it ourselves wouldn't be necessary.

bobozaur commented on messages2/src/decorators/thread.rs:
Done.

bobozaur commented on messages2/src/decorators/please_ack.rs:
```diff
@@ -0,0 +1,42 @@
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
```
Done

bobozaur commented on messages2/src/decorators/timing.rs:
Done.

bobozaur commented on messages2/src/decorators/attachment.rs:
```diff
@@ -0,0 +1,211 @@
+    pub fn new(content: AttachmentType) -> Self {
+        Self {
+            jws: None,
+            sha256: None,
```
Yeah. And the follow-up would be - what to do if it doesn't match the one in the message? Should we fail deserialization altogether? Should we have some sort of validation post-deserialization? But this is indeed for another issue 😄 .

Patrik-Stas replied:
Maybe that behaviour can be configurable

bobozaur commented on messages2/src/msg_types/types/present_proof.rs:
```diff
@@ -0,0 +1,92 @@
+#[msg_type(major = 1)]
+pub enum PresentProofV1 {
+    #[msg_type(minor = 0, roles = "Role::Prover, Role::Verifier")]
+    V1_0(PhantomData<fn() -> PresentProofV1_0>),
```
FYI, `PhantomData` is replaced with `MsgKindType` (which uses `PhantomData` internally). It's to make it easier on the eyes + provide the `kind_from_str()` method.

mirgee reviewed (dismissed):
Just minor nitpicks otherwise LGTM 👍

gmulhearn commented on messages2/src/decorators/thread.rs:
```diff
@@ -0,0 +1,96 @@
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
```
Maybe i'm missing something, but should these well-known goal codes be serde renamed so that they serialize/deserialize correctly for the RFC format? e.g. `aries.vc` etc; https://github.com/hyperledger/aries-rfcs/blob/main/concepts/0519-goal-codes/README.md#well-known-goal-codes

bobozaur replied:
Yeah, so... This used to have a variant `Other(String)` which would accept other variants, but removed it in favor of the `MaybeKnown` wrapper because there were a lot of constants involved and boilerplate code. And, of course, I forgot to add the renames to keep everything in check. I'll have another look at types like this to make sure they're implemented properly. 

Great catch, George!

bobozaur replied:
Fixed.

bobozaur commented on messages2/src/misc/mod.rs:
```diff
@@ -0,0 +1,111 @@
+        test_serde(msg, json);
+    }
+
+    pub fn test_serde<T>(value: T, json: Value)
```
Fair. Will rename.

bobozaur commented on messages_macros/src/lib.rs:
```diff
@@ -0,0 +1,63 @@
+/// a protocol or a major version encapsulating enum.
+///
+/// The minor versions are represented by the major version enum's variants
+/// and the field encapsulated in the variants are expected to be [`crate::msg_types::MsgKindType<T>`].
```
Yeah, I know. Maybe it would be better to just use the type name alone instead of the absolute path. Just wanted to have the type name there to be more descriptive. Will fix.

bobozaur replied:
Excluded the path, should be more readable now.

bobozaur commented on messages2/src/msg_fields/protocols/connection/problem_report.rs:
```diff
@@ -0,0 +1,99 @@
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+pub struct ProblemReportDecorators {
+    #[serde(rename = "~thread")]
+    pub thread: Thread,
+    #[serde(rename = "~l10n")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub localization: Option<MsgLocalization>,
+    #[serde(rename = "~timing")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub timing: Option<Timing>,
```
As discussed in our call, we're not really sure if we wanna lock into some specific public API that we expose so we kinda expose everything in terms of these messages.

bobozaur commented on messages2/src/msg_types/mod.rs:
```diff
@@ -0,0 +1,148 @@
+/// *idiomatic*.
+///
+/// Good reads and references:
+/// - https://doc.rust-lang.org/std/marker/struct.PhantomData.html
```
Links should get rendered properly now.
